crates/ra_syntax/src/validation.rs

```diff
@@ -111,55 +111,46 @@ pub(crate) fn validate(root: &SyntaxNode) -> Vec<SyntaxError> {
     errors
 }
 
-// FIXME: kill duplication
 fn validate_literal(literal: ast::Literal, acc: &mut Vec<SyntaxError>) {
+    fn unquote(text: &str, prefix_len: usize, end_delimiter: char) -> Option<&str> {
+        text.rfind(end_delimiter).and_then(|end| text.get(prefix_len..end))
+    }
+
     let token = literal.token();
     let text = token.text().as_str();
+
+    let mut push_err = |prefix_len, (off, err): (usize, unescape::EscapeError)| {
+        let off = token.text_range().start() + TextUnit::from_usize(off + prefix_len);
+        acc.push(SyntaxError::new(err.into(), off));
+    };
+
     match token.kind() {
         BYTE => {
-            if let Some(end) = text.rfind('\'') {
-                if let Some(without_quotes) = text.get(2..end) {
-                    if let Err((off, err)) = unescape::unescape_byte(without_quotes) {
-                        let off = token.text_range().start() + TextUnit::from_usize(off + 2);
-                        acc.push(SyntaxError::new(err.into(), off))
-                    }
-                }
+            if let Some(Err(e)) = unquote(text, 2, '\'').map(unescape::unescape_byte) {
+                push_err(2, e);
             }
```

Comment:
This is quite good already, but would it make sense to pull it out into a seperate function:
```rust
fn validate_char(token: &SyntaxToken, prefix_len: usize, end_delimiter: char, acc: &mut Vec<SyntaxError>) -> {
    let text = token.text().as_str();
    if let Some(Err(e)) = unquote(text, 2, '\'').map(unescape::unescape_byte) {
        let off = token.text_range().start() + TextUnit::from_usize(off + prefix_len);
        acc.push(SyntaxError::new(err.into(), off));
    }
}
```
and then the match case just becomes:
```rust
BYTE => validate_char(&token, 2, '\'', acc),
```
This would avoid the the `push_err` lambda  I wouldn't minds duplicating that part, and it would play nicely with #540.

Reply:
Yeah, only one addition, `unescape::unescape_byte()` should also be forwarded as a callback function to `validate_char()` this way the function will become generic and will have 5 parameters. If you are okay with that, tho?

Reply:
Right, hadn't noticed they used different functions, but that makes sense :) This might require a type parameter for the function, are you comfortable with that or do you need a hint?

Reply:
Yes, I am comfortable with that, just wanted our approval)